solr/core/src/java/org/apache/solr/schema/DenseVectorField.java

```diff
@@ -205,53 +254,194 @@ public IndexableField createField(SchemaField field, Object parsedVector) {
    * org.apache.solr.handler.loader.CSVLoader} produces an ArrayList of String - {@link
    * org.apache.solr.handler.loader.JsonLoader} produces an ArrayList of Double - {@link
    * org.apache.solr.handler.loader.JavabinLoader} produces an ArrayList of Float
-   *
-   * @param inputValue - An {@link ArrayList} containing the elements of the vector
-   * @return the vector parsed
    */
-  float[] parseVector(Object inputValue) {
-    if (!(inputValue instanceof List)) {
-      throw new SolrException(
-          SolrException.ErrorCode.BAD_REQUEST,
-          "incorrect vector format."
-              + " The expected format is an array :'[f1,f2..f3]' where each element f is a float");
+  public VectorBuilder getVectorBuilder(Object inputValue, VectorBuilder.BuilderPhase phase) {
+    switch (vectorEncoding) {
+      case FLOAT32:
+        return new VectorBuilder.Float32VectorBuilder(dimension, inputValue, phase);
+      case BYTE:
+        return new VectorBuilder.ByteVectorBuilder(dimension, inputValue, phase);
+      default:
+        throw new SolrException(
+            SolrException.ErrorCode.SERVER_ERROR,
+            "Unexpected state. Vector Encoding: " + vectorEncoding);
     }
-    List<?> inputVector = (List<?>) inputValue;
-    if (inputVector.size() != dimension) {
-      throw new SolrException(
-          SolrException.ErrorCode.BAD_REQUEST,
-          "incorrect vector dimension."
-              + " The vector value has size "
-              + inputVector.size()
-              + " while it is expected a vector with size "
-              + dimension);
+  }
+
+  abstract static class VectorBuilder {
```

Comment:
to discuss:  could this be an external class? maybe an util one? 
Is it more a "VectorParser" rather than a builder maybe?